File: src/ru/nsu/ccfit/zuev/osu/beatmap/parser/BeatmapParser.java
package ru.nsu.ccfit.zuev.osu.beatmap.parser;

import android.util.Log;

import com.rian.difficultycalculator.beatmap.hitobject.HitObject;
import com.rian.difficultycalculator.utils.HitObjectStackEvaluator;

import java.io.File;
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import okio.BufferedSource;
import okio.Okio;
import ru.nsu.ccfit.zuev.osu.ToastLogger;
import ru.nsu.ccfit.zuev.osu.Utils;
import ru.nsu.ccfit.zuev.osu.beatmap.BeatmapData;
import ru.nsu.ccfit.zuev.osu.beatmap.constants.BeatmapSection;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapColorParser;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapControlPointsParser;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapDifficultyParser;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapEventsParser;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapGeneralParser;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapHitObjectsParser;
import ru.nsu.ccfit.zuev.osu.beatmap.parser.sections.BeatmapMetadataParser;
import ru.nsu.ccfit.zuev.osu.helper.FileUtils;
import ru.nsu.ccfit.zuev.osu.helper.StringTable;
import ru.nsu.ccfit.zuev.osuplus.R;

/**
 * A parser for parsing <code>.osu</code> files.
 */
public class BeatmapParser {
    /**
     * The <code>.osu</code> file.
     */
    private final File file;

    /**
     * The <code>BufferedSource</code> responsible for reading the beatmap file's contents.
     */
    private BufferedSource source;

    /**
     * The format version of the beatmap.
     */
    private int beatmapFormatVersion = 14;

    /**
     * @param file The <code>.osu</code> file.
     */
    public BeatmapParser(final File file) {
        this.file = file;
    }

    /**
     * @param path The path to the <code>.osu</code> file.
     */
    public BeatmapParser(final String path) {
        file = new File(path);
    }

    private static final BeatmapGeneralParser generalParser = new BeatmapGeneralParser();
    private static final BeatmapMetadataParser metadataParser = new BeatmapMetadataParser();
    private static final BeatmapDifficultyParser difficultyParser = new BeatmapDifficultyParser();
    private static final BeatmapEventsParser eventsParser = new BeatmapEventsParser();
    private static final BeatmapControlPointsParser controlPointsParser = new BeatmapControlPointsParser();
    private static final BeatmapColorParser colorParser = new BeatmapColorParser();
    private static final BeatmapHitObjectsParser hitObjectsParser = new BeatmapHitObjectsParser();

    /**
     * Attempts to open the beatmap file.
     *
     * @return Whether the beatmap file was successfully opened.
     */
    public boolean openFile() {
        try {
            source = Okio.buffer(Okio.source(file));
        } catch (final IOException e) {
            Log.e("BeatmapParser.openFile", e.getMessage());
            source = null;
            return false;
        }

        try {
            String head = source.readUtf8Line();
            if (head == null) {
                closeSource();
                return false;
            }

            Pattern pattern = Pattern.compile("osu file format v(\\d+)");
            Matcher matcher = pattern.matcher(head);
            if (!matcher.find()) {
                closeSource();
                return false;
            }

            int formatPos = head.indexOf("file format v");
            beatmapFormatVersion = Utils.tryParseInt(head.substring(formatPos + 13), beatmapFormatVersion);
        } catch (Exception e) {
            Log.e("BeatmapParser.openFile", e.getMessage());
        }

        return true;
    }

    /**
     * Parses the <code>.osu</code> file.
     *
     * @param withHitObjects Whether to parse hit objects. This will improve parsing time significantly.
     * @return A <code>BeatmapData</code> containing relevant information of the beatmap file,
     * <code>null</code> if the beatmap file cannot be opened or a line could not be parsed.
     */
    public BeatmapData parse(boolean withHitObjects) {
        String fileName = file.getName().substring(0, file.getName().length() - 4);

        if (source == null && !openFile()) {
            ToastLogger.showText(
                    StringTable.format(R.string.beatmap_parser_cannot_open_file, fileName),
                    true
            );
            return null;
        }

        BeatmapSection currentSection = null;
        BeatmapData data = new BeatmapData();

        data.setMD5(FileUtils.getMD5Checksum(file));
        data.setFolder(file.getParent());
        data.setFilename(file.getPath());
        data.setFormatVersion(beatmapFormatVersion);

        String s;

        try {
            while ((s = source.readUtf8Line()) != null) {
                // Check if beatmap is not an osu!standard beatmap
                if (data.general.mode != 0) {
                    // Silently ignore (do not log anything to the user)
                    return null;
                }

                // Handle space comments
                if (s.startsWith(" ") || s.startsWith("_")) {
                    continue;
                }

                // Now that we've handled space comments, we can trim space
                s = s.trim();

                // Handle C++ style comments and empty lines
                if (s.startsWith("//") || s.isEmpty()) {
                    continue;
                }

                // [SectionName]
                if (s.startsWith("[") && s.endsWith("]")) {
                    currentSection = BeatmapSection.parse(s.substring(1, s.length() - 1));
                    continue;
                }

                if (currentSection == null) {
                    continue;
                }

                try {
                    switch (currentSection) {
                        case general:
                            generalParser.parse(data, s);
                            break;
                        case metadata:
                            metadataParser.parse(data, s);
                            break;
                        case difficulty:
                            difficultyParser.parse(data, s);
                            break;
                        case events:
                            eventsParser.parse(data, s);
                            break;
                        case timingPoints:
                            controlPointsParser.parse(data, s);
                            break;
                        case colors:
                            colorParser.parse(data, s);
                            break;
                        case hitObjects:
                            if (withHitObjects) {
                                hitObjectsParser.parse(data, s);
                            }
                            break;
                    }
                } catch (Exception e) {
                    Log.e("BeatmapParser.parse", "Unable to parse line " + s, e);
                }
            }

            closeSource();
            populateObjectData(data);
        } catch (IOException e) {
            Log.e("BeatmapParser.parse", e.getMessage());
            return null;
        }

        return data;
    }

    /**
     * Populates the object scales of a <code>BeatmapData</code>.
     *
     * @param data The <code>BeatmapData</code> whose object scales will be populated.
     */
    public static void populateObjectData(final BeatmapData data) {
        float scale = (1 - 0.7f * (data.difficulty.cs - 5) / 5) / 2;

        for (HitObject object : data.hitObjects.getObjects()) {
            object.setScale(scale);

            // Reset stack height as we will be re-applying it.
            object.setStackHeight(0);
        }

        HitObjectStackEvaluator.applyStacking(data.getFormatVersion(), data.hitObjects.getObjects(), data.difficulty.ar, data.general.stackLeniency);
    }

    private void closeSource() {
        if (source != null) {
            try {
                source.close();
            } catch (IOException ignored) {}

            source = null;
        }
    }
}

File: src/com/rian/osu/beatmap/parser/BeatmapParser.kt
package com.rian.osu.beatmap.parser

import android.util.Log
import com.rian.osu.GameMode
import com.rian.osu.beatmap.Beatmap
import com.rian.osu.beatmap.BeatmapProcessor
import com.rian.osu.beatmap.constants.BeatmapSection
import com.rian.osu.beatmap.parser.sections.*
import okio.BufferedSource
import okio.buffer
import okio.source
import ru.nsu.ccfit.zuev.osu.helper.FileUtils
import java.io.File
import java.io.IOException
import java.util.regex.Pattern
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.ensureActive

/**
 * A parser for parsing `.osu` files.
 */
class BeatmapParser {
    /**
     * The `.osu` file.
     */
    private val file: File

    /**
     * The [CoroutineScope] to use for coroutines.
     */
    private val scope: CoroutineScope?

    /**
     * The precomputed MD5 hash of the beatmap file, if available.
     *
     * This is used to avoid unnecessary MD5 calculations when parsing.
     */
    private val precomputedMD5: String?

    /**
     * @param file The `.osu` file.
     * @param scope The [CoroutineScope] to use for coroutines.
     */
    @JvmOverloads
    constructor(file: File, scope: CoroutineScope? = null, precomputedMD5: String? = null) {
        this.file = file
        this.scope = scope
        this.precomputedMD5 = precomputedMD5
    }

    /**
     * @param path The path to the `.osu` file.
     * @param scope The [CoroutineScope] to use for coroutines.
     */
    @JvmOverloads
    constructor(path: String, scope: CoroutineScope? = null, precomputedMD5: String? = null) : this(File(path), scope, precomputedMD5)

    /**
     * Parses the `.osu` file.
     *
     * @param withHitObjects Whether to parse hit objects. Setting this to `false` will improve parsing time significantly.
     * @param mode The [GameMode] to parse for. Defaults to [GameMode.Standard].
     * @return A [Beatmap] containing relevant information of the beatmap file.
     * @throws IOException If an I/O error occurs while reading the file.
     * @throws NumberFormatException If the beatmap's file version cannot be determined.
     * @throws IllegalArgumentException If the beatmap is not an osu!standard beatmap.
     */
    @JvmOverloads
    @Throws(IOException::class, IllegalArgumentException::class, NumberFormatException::class)
    fun parse(withHitObjects: Boolean, mode: GameMode = GameMode.Standard) = openFile().use { source ->
        scope?.ensureActive()

        // Check for format version first to avoid unnecessary MD5 calculation.
        val formatVersion = getFormatVersion(source)

        var currentLine: String?
        var currentSection: BeatmapSection? = null

        val beatmap = Beatmap(mode).also {
            it.md5 = precomputedMD5 ?: FileUtils.getMD5Checksum(file)
            it.filePath = file.path
            it.formatVersion = formatVersion
        }

        while (source.readUtf8Line().also { currentLine = it } != null) {
            scope?.ensureActive()

            if (beatmap.general.mode != 0) {
                throw IllegalArgumentException("Not an osu!standard beatmap")
            }

            var line = currentLine ?: continue

            // Handle space comments
            if (line.startsWith(" ") || line.startsWith("_")) {
                continue
            }

            // Now that we've handled space comments, we can trim space
            line = line.trim { it <= ' ' }

            // Handle C++ style comments and empty lines
            if (line.startsWith("//") || line.isEmpty()) {
                continue
            }

            // [SectionName]
            if (line.startsWith("[") && line.endsWith("]")) {
                currentSection = BeatmapSection.parse(line.substring(1, line.length - 1))

                // HitObjects are always in the last section (since it is dependent on other things such as
                // difficulty, control points, etc.)
                // We can stop here if we do not need to parse them.
                if (currentSection == BeatmapSection.HitObjects && !withHitObjects) {
                    break
                }

                continue
            }

            if (currentSection == null) {
                continue
            }

            scope?.ensureActive()

            try {
                when (currentSection) {
                    BeatmapSection.General ->
                        BeatmapGeneralParser.parse(beatmap, line, scope)

                    BeatmapSection.Metadata ->
                        BeatmapMetadataParser.parse(beatmap, line, scope)

                    BeatmapSection.Difficulty ->
                        BeatmapDifficultyParser.parse(beatmap, line, scope)

                    BeatmapSection.Events ->
                        BeatmapEventsParser.parse(beatmap, line, scope)

                    BeatmapSection.TimingPoints ->
                        BeatmapControlPointsParser.parse(beatmap, line, scope)

                    BeatmapSection.Colors ->
                        BeatmapColorParser.parse(beatmap, line, scope)

                    BeatmapSection.HitObjects ->
                        BeatmapHitObjectsParser.parse(beatmap, line, scope)

                    else -> continue
                }
            } catch (e: Exception) {
                if (e is CancellationException) {
                    throw e
                }

                Log.e("BeatmapParser.parse", "Unable to parse line", e)
            }
        }

        val processor = BeatmapProcessor(beatmap, scope)

        processor.preProcess()

        beatmap.hitObjects.objects.forEach {
            scope?.ensureActive()

            it.applyDefaults(beatmap.controlPoints, beatmap.difficulty, mode, scope)
            it.applySamples(beatmap.controlPoints, scope)
        }

        processor.postProcess()

        beatmap
    }

    private fun openFile() = file.source().buffer()

    @Throws(IOException::class, NumberFormatException::class)
    private fun getFormatVersion(source: BufferedSource): Int {
        val head = source.readUtf8Line() ?: throw IOException("Empty file")
        val pattern = Pattern.compile("osu file format v(\\d+)")
        val matcher = pattern.matcher(head)

        if (!matcher.find()) {
            throw NumberFormatException("Invalid format version")
        }

        return head.substring(head.indexOf("file format v") + 13).toIntOrNull()
            ?: throw NumberFormatException("Invalid format version")
    }
}
